sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -837,8 +837,36 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
     val j = ctx.freshName("j")
     val values = ctx.freshName("values")
 
+    val isPrimitiveFrom = ctx.isPrimitiveType(fromType)
```

Comment:
we need to make sure the input array's element nullability is false, but primitive type array doesn't guarantee it. e.g. we can have `ArrayType(ByteType, true)`